index/mem/index.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	return &memIndex{
+		opts:   opts,
+		hashFn: hashFn,
+
+		docIDGen: atomic.NewUint32(0),
+
+		idsMap:       make(map[doc.Hash]index.DocID, initialSize),
+		documents:    make(map[index.DocID]document, initialSize),
+		fieldMap:     make(map[string]*termsDict, initialSize),
+		postingsList: newPostingsList(initialSize),
+	}
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *memIndex) Insert(d doc.Document) error {
+	// early terminate if id is already present
+	docHash := i.hashFn(d.ID)
+	i.idsLock.RLock()
+	_, ok := i.idsMap[docHash]
+	i.idsLock.RUnlock()
+	if ok {
+		return nil
+	}
+
+	// insert into document id map
+	i.idsLock.Lock()
+	// check if it's been inserted since we released lock
+	_, ok = i.idsMap[docHash]
+	if ok {
+		i.idsLock.Unlock()
+		return nil
+	}
+	nextDocID := index.DocID(i.docIDGen.Inc())
+	i.idsMap[docHash] = nextDocID
+	i.idsLock.Unlock()
+
+	i.insertDocument(document{
+		Document: d,
+		docID:    nextDocID,
+	})
+	return nil
+}
+
+func (i *memIndex) insertDocument(doc document) {
+	// insert document into master doc id -> doc map
+	i.documentsLock.Lock()
+	i.documents[doc.docID] = doc
+	i.documentsLock.Unlock()
+
+	// insert each of the indexed fields into the reverse index
+	for _, field := range doc.IndexableFields {
+		i.insertField(field, doc.docID)
+	}
+}
+
+func (i *memIndex) insertField(field doc.Field, docID index.DocID) {
+	fieldName := string(field.Name)
+	// read lock to see if field key has been seen before
+	i.fieldLock.RLock()
+	fieldTerms, ok := i.fieldMap[fieldName]
+	i.fieldLock.RUnlock()
+	if ok {
+		fieldTerms.Insert(string(field.Value), docID)
+		return
+	}
+
+	// i.e. td does not exist, create it
+	i.fieldLock.Lock()
+	// has the key been inserted since the lock was last retrieved
+	fieldTerms, ok = i.fieldMap[fieldName]
+	// if not, create it
+	if !ok {
+		fieldTerms = newTermsDict(i.postingsList)
+		i.fieldMap[fieldName] = fieldTerms
+	}
+	i.fieldLock.Unlock()
+
+	// perform the insert finally
+	fieldTerms.Insert(string(field.Value), docID)
+}
+
+func (i *memIndex) Fetch(opts index.FetchOptions) []doc.Document {
+	// TODO(prateek): timeout/early termination once we know we're done
+
+	var (
+		docsSet         = false
+		candidateDocIds *roaring.Bitmap
+	)
+
+	for _, filter := range opts.Filters {
+		filter := filter
+
+		i.fieldLock.RLock()
+		fieldTerms, ok := i.fieldMap[string(filter.FieldName)]
+		i.fieldLock.RUnlock()
+
+		// i.e. we don't have any documents for the given filter, can early terminate entire fn
+		if !ok {
+			return nil
+		}
+
+		fetchedIds, ok := fieldTerms.Fetch(string(filter.FieldValueFilter), filter.Regexp, filter.Negate)
+		// i.e. we don't have any documents for the given filter, can early terminate entire fn
+		if !ok {
+			return nil
+		}
+
+		if !docsSet {
+			candidateDocIds = fetchedIds.clone()
+			docsSet = true
+			continue
+		}
+
+		// update candidate set
+		fetchedIds.andTo(candidateDocIds)
```

Comment:
Instead of intersecting one by one, perhaps its preferable to wait until we've extracted all IDs? Then we can optimize the intersection. For example, if any our empty we can just return without actually doing any intersections. Barring that, we can then perform the intersections in size order, smallest first.

Reply:
There's a bunch of ways to optimise this. For e.g. if we are applying 10 filters which eventually lead to no data, but take a long time to retrieve respective doc lists then early application can helpful in pipelining cpu cost. 

I'll create an interface where we can accumulate `fetchedIds`. Something like `docFetcher`, which'll make the decisions based on w/e heuristics we come up with.